quixstreams/processing_context.py

```diff
@@ -70,13 +73,19 @@ def commit_checkpoint(self, force: bool = False):
 
         :param force: if `True`, commit the checkpoint before its expiration deadline.
         """
-        if not self._checkpoint.empty() and (self._checkpoint.expired() or force):
-
-            logger.debug(f"Committing a checkpoint force={force}")
+        if self._checkpoint.expired() or force:
+            logger.debug(f"Attempting checkpoint commit; forced={force}")
             start = time.monotonic()
             self._checkpoint.commit()
             elapsed = round(time.monotonic() - start, 2)
             logger.debug(
-                f"Committed a checkpoint force={force} time_elapsed={elapsed}s"
+                f"Committed a checkpoint; forced={force}, time_elapsed={elapsed}s"
             )
             self.init_checkpoint()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.exactly_once:
+            self.producer.abort_transaction(5)
```

Comment:
Ok, that makes sense.

Besides, we cannot easily do it in `RowProducer` because `RowProducer.__exit__` triggers `.flush()`, which is managed in the Checkpoint when running with the `Application`.